### PR TITLE
[GPU] Skip reorder_node_to_split to avoid change of input data type for ondenn kernel support

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
 - Skip reorder_node_to_split to avoid change of input data type for ondenn kernel support

### Tickets:
 - 107131

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp [hunk 1]
```diff
@@ -75,11 +79,46 @@ void handle_reshape::run(program& p) {
...
+            // find onednn users
             for (const auto& user : node->get_users()) {
                 if (user->is_type<reorder>())
                     reorder_node_to_split.push_back(user);
+                if (user->get_preferred_impl_type() == cldnn::impl_types::onednn)
+                    onednn_users.push_back(user);
+            }
+
+            // If onednn user doesn't support new input data type, remove target reorder_node
+            if (!onednn_users.empty() && !reorder_node_to_split.empty()) {
```
Comment (jade-cho) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:94]: Could you add more comments why the siblings of reorder should be checked?

**Review by jade-cho — APPROVED** [review]
LGTM

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp (hunk 1 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:95]: Added more comments. Thanks!

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp [hunk 2]
```diff
@@ -75,11 +79,47 @@ void handle_reshape::run(program& p) {
...
+                                auto wei_dt = is_conv ? user->as<convolution>().weights().get_output_layout().data_type :
+                                                        user->as<deconvolution>().weights().get_output_layout().data_type;
+                                onednn_support = layout_optimizer::onednn_check_data_types_for_convolution(output_data_type, wei_dt, out_dt);
+                            } else if (user->is_type<pooling>()) {
+                                onednn_support = layout_optimizer::onednn_check_data_types_for_pooling(output_data_type, out_dt);
+                            }
+                        }
+                        if (!onednn_support) {
+                            reorder_node_to_split.erase(std::remove(reorder_node_to_split.begin(), reorder_node_to_split.end(), reorder_node),
+                                                        reorder_node_to_split.end());
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:118]: Is it safe to remove items from the list while you are iterating over it? I think vector may invalidate its iterators when you erase an element

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:118]: @sshlyapn When I checked it from reorder_node_to_split size=1 to size=0, there was no issue. Do you know any way to check potential issue here?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:118]: > Invalidates iterators and references at or after the point of the erase, including the [end()](https://en.cppreference.com/w/cpp/container/vector/end) iterator.

https://en.cppreference.com/w/cpp/container/vector/erase

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:120]: Modified to avoid removing items from the list while iterating over it. Thanks!

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp [hunk 3]
```diff
@@ -75,11 +79,49 @@ void handle_reshape::run(program& p) {
...
             // find the users of reshape that are reorder type, if none present then skip the current node
+            // find users who are onednn impl
             for (const auto& user : node->get_users()) {
                 if (user->is_type<reorder>())
                     reorder_node_to_split.push_back(user);
+                if (user->get_preferred_impl_type() == cldnn::impl_types::onednn)
+                    onednn_users.push_back(user);
+            }
+
+            // If onednn user doesn't support new input data type from future "reorder:_reshape_input_" reorder,
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:93]: [random spot] @wilson-seok Could you check if we still need reorder split at all? Maybe we can just remove that part and don't add this custom logic for onednn at all?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:93]: @vladimir-paramuzov I'm checking performance regression impact of removing whole reorder split. There is no regression on gen9 platform. I'm checking on gen12lp. BTW you created new PR which touch reorder split in handle_reshape(https://github.com/openvinotoolkit/openvino/pull/16911). Do you think we will keep it? Or remove if there is no performance regression?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:93]: If reorder split won't impact anything and can be removed, then handle_reshape modification in my PR can be discarded as well.

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:93]: Just finished performance regression check on gen12lp. There is no critical performance regression on both gen9 and gel12lp. I will update this PR to remove it. Thanks!

Comment (isanghao) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp]: Then I think it is better to let @jade-cho  to finalize his code. Partial implementation from this PR doesn't seem to be necessary. (especially because Jade's work is not complete yet)

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp]: @wilson-seok I guess accuracy should be checked in the first place, as this code seem to be designed to avoid some issues with buffers propagation for optimized out primitives

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp]: > @wilson-seok I guess accuracy should be checked in the first place, as this code seem to be designed to avoid some issues with buffers propagation for optimized out primitives

Let me try accuracy check for this PR.

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp]: @vladimir-paramuzov Removing reorder split causes exec graph diff in 82 models. Accuracy check for them should take too long time. So why don't we keep it and add custom condition for onednn?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:93]: Reverted to previous fix. It is hard to check accuracy impact within CF because 82 models are affected. Will create a ticket in backlog.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 4]
```diff
@@ -263,8 +304,21 @@ bool layout_optimizer::can_fuse_reorder(program_node& prev, program_node& next,
...
+            return false;
+        }
+        return true;
+    }
+
+    // Target patterns
+    // - Fuse reorder into Permute: any node -> reorder -> Permute => any node -> Permute
+    // - Fuse reorder into FC     : any node -> reorder -> FC      => any node -> FC
+    if ((next.is_type<permute>() || next.is_type<fully_connected>()) && (fmt_prev == fmt_next) &&
+        ((fmt_prev != format::any) && (fmt_next != format::any))) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:319]: this is not related to this PR. please remove this change.

File: src/plugins/intel_gpu/tests/passes/handle_reshape.cpp [hunk 5]
```diff
@@ -50,3 +50,42 @@ TEST(handle_reshape, dont_remove_reshape_that_changes_rank) {
...
+    lo.set_optimization_attribute(layout_optimizer::optimization_attributes_type::use_onednn_impls, true);
+    reorder_factory rf;
+
+    program_wrapper::apply_opt_pass<reorder_inputs>(*prog, lo, rf);
+    program_wrapper::apply_opt_pass<handle_reshape>(*prog);
+
+    ASSERT_NE(prog, nullptr);
+
+    ASSERT_TRUE(prog->get_node("matmul").get_dependency(0).get_output_layout().data_type == data_types::f16);
+}
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/passes/handle_reshape.cpp:91]: please add an empty line at the end of the file

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/tests/passes/handle_reshape.cpp (hunk 5 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/passes/handle_reshape.cpp]: Fixed. Thanks!

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 6]
```diff
@@ -451,15 +505,33 @@ bool layout_optimizer::can_fuse_reorder_to_prev(program_node& prev, reorder_node
...
 
         if (prev.is_type<eltwise>() &&
             is_mixed_layout(prev, *next, false, {{ format::bs_fs_zyx_bsv32_fsv32, format::bs_fs_zyx_bsv32_fsv16 }}))
             return true;
     }
 
+    if (prev.is_type<eltwise>() && (fmt_prev == fmt_next)) {
+        return true;
+    }
+
```
[the remark this answers is not in the document]
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Removed. Thanks!

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 4 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Removed. Thanks!

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 7]
```diff
@@ -451,9 +492,23 @@ bool layout_optimizer::can_fuse_reorder_to_prev(program_node& prev, reorder_node
...
+            } else {
+                if (prev.is_type<fully_connected>()) {
+                    auto src_dtype = prev.as<fully_connected>().input().get_output_layout().data_type;
+                    auto weight_dtype = prev.as<fully_connected>().weights().get_output_layout().data_type;
+                    // Check data type combination for oneDNN inner product
+                    if (onednn_check_data_types_for_fc_gemm(src_dtype, weight_dtype, reorder_dtype))
+                        return true;
+                }
+            }
+        }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:511]: is it intended change from your PR?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Removed. Thanks!

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp [hunk 8]
```diff
@@ -75,11 +79,49 @@ void handle_reshape::run(program& p) {
...
+            // remove target reorder_node to keep original datatype
+            if (!onednn_users.empty() && !reorder_node_to_split.empty()) {
+                // Copy reorder_node_to_split to iteration
+                std::vector<program_node*> reorder_users(reorder_node_to_split);
+                for (const auto& reorder_node : reorder_users) {
+                    auto output_data_type = reorder_node->get_output_layout().data_type;
+                    bool onednn_support = true;
+                    for (const auto& user : onednn_users) {
+                        if (user->get_preferred_impl_type() == cldnn::impl_types::onednn) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp:102]: (no need to apply in this PR) i think this line can be removed.

File: src/plugins/intel_gpu/tests/passes/handle_reshape.cpp [hunk 9]
```diff
@@ -50,3 +50,42 @@ TEST(handle_reshape, dont_remove_reshape_that_changes_rank) {
 
     ASSERT_TRUE(prog->get_node("reshape").can_be_optimized());
 }
+
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/passes/handle_reshape.cpp:53]: (no need to apply in this PR) Please add a comment: '''// Onednn FC does not support fp32 input, fp16 weight. In such case, we need to ignore reorder_split from handle_reshape pass'''

**Review by isanghao — APPROVED** [review]

File: src/plugins/intel_gpu/tests/passes/handle_reshape.cpp (hunk 9 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/passes/handle_reshape.cpp:53]: Updated. Thanks!

File: src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp (hunk 8 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/handle_reshape.cpp]: Removed. Thanks!
